### DISCUSSION TITLE
My twinBASIC DLL Experiment

### DISCUSSION THREAD

**Discussion by GCuser99** [GCuser99/SeleniumVBA#59] [Ideas]
This is my report on attempting to convert SeleniumVBA to an Office ActiveX DLL using the new (currently in Beta) [twinBASIC](https://twinbasic.com/preview.html).

TwinBasic is a VBA/VB6 development environment that promises to bring VB6 into the world of Win64 architecture, from where MS abandoned it many years ago as a 32-bit programming language, with the inception of VB.Net. In addition to aiming for VBx backwards compatibility, twinBASIC extends the language to include many new features such as parameter overloading, new statements, assignment operators, and data types. The author's name is Wayne Phillips, who seems to working tirelessly to improve the compiler based on quite a lot of user feedback, including mine.

One significant advantage of twinBASIC is that it "generates very small executables that do not require any runtime library or redistributable components", so it's pretty easy to install the generated ActiveX DLL locally.

I will provide a 64-bit compiled DLL, as well as the tB project files in an experimental folder of the SeleniumVBA GitHub repository in case anyone reading this would like to give it a try. Be aware that there is a 5-second twinBASIC banner that will appear once during each SeleniumVBA session for non-paid subscribers. The IDE can be downloaded from [here](https://github.com/twinbasic/twinbasic/releases).

For the most part, the SeleniumVBA source code was compiled without issue. I submitted about a half dozen error reports of which most were fixed promptly. There were, however, a couple of twinBASIC errors (or features?) that required workarounds. Additionally, there were a few changes required to reflect the differences between running using the VBA interpreter versus from a compiled ActiveX DLL. I will review most of the changes below. These changes were exclusively applied in my twinBASIC project - NOT to the VBA Add-in source.

**The Application Object.**
SeleniumVBA requires a reference to the Office host application in order to calculate relative paths. The default base path is to the active VBA calling project's host document folder, which can be different from the path where the SeleniumVBA source resides. For the DLL to know the path of the calling document, the Office Application object either needs to be passed to the DLL by the caller, or the DLL would have to be capable of detecting the caller's Application object. After a lot of research, I decided on a solution to the latter, which I admit seems a little dicey, but so far has worked flawlessly with my testing to date. The technique below determines the calling process ID (PID), and then searches the open windows tree structure for an Office window signature belonging to that PID that can be converted to the Application object via the AccessibleObjectFromWindow api.

Here is the code for CallerApplicationObject:

EDIT: This discussion, specifically about the code below, was referenced by [Mike Wolfe in the DevCon 2024](https://nolongerset.com/devcon-2024-tb/). With that potential increased attention, I've made some refinements to improve/simplify the code logic. I have also made the code compatible with [Jon Johnson's Windows Development Library](https://github.com/fafalone/WinDevLib?tab=readme-ov-file#windevlib).

```vba
Attribute VB_Name = "tbUtils"
' ==========================================================================
' SeleniumVBA v5.3
'
' A Selenium wrapper for browser automation developed for MS Office VBA
'
' https://github.com/GCuser99/SeleniumVBA/tree/main
'
' Contact Info:
'
' https://github.com/6DiegoDiego9
' https://github.com/GCUser99
' ==========================================================================
' MIT License
'
' Copyright (c) 2023-2024, GCUser99 and 6DiegoDiego9 (https://github.com/GCuser99/SeleniumVBA)
'
' Permission is hereby granted, free of charge, to any person obtaining a copy
' of this software and associated documentation files (the "Software"), to deal
' in the Software without restriction, including without limitation the rights
' to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
' copies of the Software, and to permit persons to whom the Software is
' furnished to do so, subject to the following conditions:
'
' The above copyright notice and this permission notice shall be included in all
' copies or substantial portions of the Software.
'
' THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
' IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
' FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
' AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
' LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
' OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
' SOFTWARE.
' ==========================================================================

Option Private Module
Option Explicit

'Windows Dev Library Declarations (not needed if using reference to tB's WDL Package)
/*
Private Declare PtrSafe Function GetCurrentProcessId Lib "kernel32" () As Long
Private Declare PtrSafe Function GetWindowThreadProcessId Lib "user32" (ByVal hwnd As LongPtr, Optional lpdwProcessId As Long) As Long
Private DeclareWide PtrSafe Function FindWindowEx Lib "user32" Alias "FindWindowExW" (ByVal hwndParent As LongPtr, ByVal hWndChildAfter As LongPtr, ByVal lpszClass As String, ByVal lpszWindow As String) As LongPtr
Private DeclareWide PtrSafe Function IIDFromString Lib "ole32" (ByVal lpsz As String, lpiid As UUID) As Long
Private Declare PtrSafe Function AccessibleObjectFromWindow Lib "oleacc" (ByVal hwnd As LongPtr, ByVal dwId As Long, riid As UUID, ppvObject As Any) As Long

Private Const IIDSTR_IDispatch As String = "{00020400-0000-0000-C000-000000000046}"
Private Const OBJID_NATIVEOM As Long = &HFFFFFFF0

Private Type UUID
    Data1 As Long
    Data2 As Integer
    Data3 As Integer
    Data4(7) As Byte
End Type
*/

'adapted from:
'http://exceldevelopmentplatform.blogspot.com/2019/01/vba-code-to-get-excel-word-powerpoint.html
'https://codereview.stackexchange.com/questions/268498/collecting-all-instances-of-an-office-application-excel-powerpoint-word
    
Public Function callerApplicationObject() As Object
    Dim pid As Long
    Dim obj As Object

    'get the calling application process id
    pid = GetCurrentProcessId()

    'Check if we have an Excel application
    Set obj = findAO(pid, Array("XLMAIN", "XLDESK", "EXCEL7"))
    If Not obj Is Nothing Then
        Set callerApplicationObject = obj
        Exit Function
    End If
    
    'Check if we have an Access application
    Set obj = findAO(pid, Array("OMAIN"))
    If Not obj Is Nothing Then
        Set callerApplicationObject = obj
        Exit Function
    End If
    
    /* Our app is not designed to work in PPT and Word so we don't need to check for below
    'Check if we have an PowerPoint application
    Set obj = findAO(pid, Array("PPTFrameClass", "MDIClient", "mdiClass"))
    If Not obj Is Nothing Then
        Set callerApplicationObject = obj
        Exit Function
    End If

    'Check if we have an Word application
    Set obj = findAO(pid, Array("OpusApp", "_WwF", "_WwB", "_WwG"))
    If Not obj Is Nothing Then
        Set callerApplicationObject = obj
        Exit Function
    End If
    */
End Function

Private Function findAO(ByVal targetPid As Long, winTree As Variant) As Object
    Dim obj As Object
    Dim hwndMain As LongPtr
    Dim hwndChild As LongPtr
    Dim hwndParent As LongPtr
    Dim i As Long
    Dim iid As UUID
    Dim thisPid As Long
    Dim className As String
    Dim foundIt As Boolean

    className = winTree(LBound(winTree)) 'top level class name
    
    'find top-level window that matches current PID
    Do While TryGetNextWindowHwnd(className, hwndMain, outNextHwnd:= hwndMain)
        Call GetWindowThreadProcessId(hwndMain, thisPid)
        If thisPid = targetPid Then foundIt = True: Exit Do
    Loop
    
    If Not foundIt Then Exit Function
    
    'now drill down levels to find the handle of the application window
    hwndParent = hwndMain
    'in case of ms access, the "tree" has already been processed (only one level)
    hwndChild = hwndMain
    For i = LBound(winTree) + 1 To UBound(winTree)
        className = winTree(i)
        hwndChild = FindWindowEx(hwndParent, 0&, className, vbNullString)
        hwndParent = hwndChild
    Next i
    
    'construct iid that will be used in call to AccessibleObjectFromWindow
    Call IIDFromString(IIDSTR_IDispatch, iid)
            
    If AccessibleObjectFromWindow(hwndChild, OBJID_NATIVEOM, iid, obj) = 0& Then
        Set findAO = obj.Application
    End If
End Function

'from Greedo's review in second link shown above
Private Function TryGetNextWindowHwnd(ByVal className As String, ByVal prevHWnd As LongPtr, ByRef outNextHWnd As LongPtr) As Boolean
    outNextHWnd = FindWindowEx(0&, prevHWnd, className, vbNullString)
    TryGetNextWindowHwnd = outNextHWnd <> 0&
End Function

```
**Description Annotations.**
twinBASIC recognizes and uses the Description annotations from RubberDuck to add member descriptions that can be seen in VBA's Object Browser. However, it interprets these as _additions_ to the VB_Description Attribute values, thus resulting in double posting of the descriptions (I have reported this [here](https://github.com/twinbasic/twinbasic/issues/1415)). To work-around this behavior, I replaced all "[at symbol]Description" text with "%Description" in the twinBASIC project as in below.
```vba
'%Description("Clicks the element")
Public Sub Click(Optional element As WebElement)
Attribute Click.VB_Description = "Clicks the element"
    Dim data As New Dictionary
    If Not element Is Nothing Then
        data.Add "id", element.elementId
        Execute tCMD.CMD_CLICK_ELEMENT, data
    Else
        Execute tCMD.CMD_CLICK, data
    End If
End Sub
```
**Hiding NewEnum Enumerator from [IntelliSense](https://learn.microsoft.com/en-us/visualstudio/ide/using-intellisense).**
It cannot be done with VBA source, but twinBasic allows the VB6 syntax of adding a leading underscore to hide a member from Intellisense.
```vba
'@Enumerator
Public Function _NewEnum() As IUnknown
Attribute _NewEnum.VB_UserMemId = -4
    Set _NewEnum = colWebElement.[_NewEnum]
End Function

```
**Path to Code Library.**
With the addition of the optional settings file function in SeleniumVBA, the DLL needs an equivalent to Excel's Application.ThisWorkbook.Path in order to locate the SeleniumVBA.ini file. Fortunately, VB6 has the "app" object that has a member "Path" that returns the needed path:
```vba
Public Function ThisLibFolderPath() As String
    ThisLibFolderPath = App.Path
End Function
```
**The VBA Debug Object.**
All Debug.Prints in our code need to be commented out as the DLL is not aware of the Debug object, nor the VBA Immediate window.

**VBA Constants.**
There are some issues with referencing built-in constants using the VBA qualifier, as I have reported [here](https://github.com/twinbasic/twinbasic/issues/1395#issue-1474184517) and [here](https://github.com/twinbasic/twinbasic/issues/1396#issue-1474187499). So minor changes need to be made as in the following: 

```vba
'In WebDriver class must replace VBA.ColorConstants by either ColorConstants or null string
'Yellow = VBA.ColorConstants.vbYellow 'is not recognized
Yellow = ColorConstants.vbYellow  'this works
Yellow = vbYellow  'and so does this
```
**Escaping Description Characters.**
The description attributes for the KeyBoard class contained the "\\" character in the text that needed to be escaped for VBA, but not needed for twinBASIC:
```vba
'VBA description requires an escaped "\" character in the text
Attribute NullKey.VB_Description = "Returns the Null (\\ue000) key"

'twinBasic does not
Attribute NullKey.VB_Description = "Returns the Null (\ue000) key"
```
**Results!!!**
I was able to run all of the SeleniumVBA test modules using the referenced DLL without problems, once the reported bugs were fixed by Wayne, and workarounds implemented by myself. I've been using the DLL now in my own code and so far, after several weeks, no issues. I will include in the repository Excel and Access documents with the test modules in case you wish to try using the DLL. All that you will have to do is register (see below) and then reference the DLL in your VBA projects.

**Performance.**
As a compiled DLL, it is expected to be more cpu-efficient than when running in the VBA interpreter. In fact, that seems to be the case. The following code snippet consumes about 50% less execution time using the DLL as compared with running the same code in the Add-in.
```vba
'on my setup this takes 2.4 secs using the DLL, versus 4.5 secs using the Excel Add-in (21 element table)
for i = 1 to 100
    v = driver.FindElement(By.ID, "mytable").TableToArray()
next i
```
**ActiveX DLL Registration.**
In order for your Office application to set a reference to the ActiveX DLL and use it, the DLL must be registered. This is a potential hornet's nest, especially for me since I have little Windows registry expertise, especially as it concerns ActiveX components. If working from the twinBASIC compiler, registration is handled automatically every time the project is compiled. twinBASIC defaults to registering the DLL to the HKEY_CURRENT_USER hive, although that can optionally be changed to HKEY_LOCAL_MACHINE if user has admin permissions to do that.

Alternatively, if not recompiling from scratch using twinBASIC compiler, the DLL can be copied to another machine or user account and registered via regserver32. As I understand it, best practice when upgrading an already registered DLL is to first unregister, then replace the DLL with the newer version, and then register the new version:

From Command terminal prompt:
1) cd to the location of the already registered DLL.
2) Enter the following command to unregister %systemroot%\System32\regsvr32.exe /u SeleniumVBA_win64.dll
3) Now replace the old DLL with the new upgraded version.
4) Enter the following command to register %systemroot%\System32\regsvr32.exe SeleniumVBA_win64.dll

**Future Plans.**
I intend to maintain the [ActiveX DLL version of SeleniumVBA on our GitHub repository](https://github.com/GCuser99/SeleniumVBA/tree/main/dev/twinBasic) at least for a while, if not indefinitely. I have written code to make the modifications above and import to twinBASIC for easier maintenance, without affecting the SeleniumVBA Add-in source. However, as my time allows, I might only do this for significant SeleniumVBA version upgrades.

Some possible future sub-projects:
1) ~~Look into possible install package (like Inno Setup) to make registration and install more straight forward.~~ DONE! [see here](https://github.com/GCuser99/SeleniumVBA/tree/main/dist)
2) ~~Look into feasibility of creating a usable standard DLL to avoid registration.~~ DONE! [see here](https://github.com/GCuser99/SeleniumVBA/discussions/73)
3) Keep an eye out for how new language extensions could improve SeleniumVBA useability.

**For More Information.**
[twinBASIC website](https://twinbasic.com/preview.html)
[twinBASIC Releases GitHub](https://github.com/twinbasic/twinbasic/releases)
[Mike Wolfe's NoLongerSet twinBASIC Tutorials](https://nolongerset.com/tag/twinbasic/)
[Mike Wolfe's NoLongerSet twinBASIC Weekly Update](https://nolongerset.com/tag/twinbasic-weekly-update/)

If this looks like a viable solution for you, then let us know in this discussion how it goes and share your feedback!

**Comment by MarkJohnstoneGitHub** [comment]
Some  handy links regarding ActiveX Dll Registration
The stack overflow I think has a  procedure for Dll Registration without registration I haven't got around to  try out yet.  

http://www.geeksengine.com/article/register-dll.html
https://stackoverflow.com/questions/37074533/how-to-call-net-methods-from-excel-vba

**Reply by 6DiegoDiego9** [reply]
I used the mscorlib/mscoree solution for SeleniumBasic some years ago.
If I remember well, it works only for .NET assemblies and stopped to work on my PC after I moved to 64bit.
I didn't delve into it much, though...
